crates/bevy_ecs/src/observer/runner.rs

```diff
@@ -16,21 +17,30 @@ use bevy_ptr::PtrMut;
 ///
 /// Typically refers to the default runner that runs the system stored in the associated [`Observer`] component,
 /// but can be overridden for custom behavior.
-pub type ObserverRunner = fn(DeferredWorld, ObserverTrigger, PtrMut, propagate: &mut bool);
+///
+/// See `observer_system_runner` for safety considerations.
+pub type ObserverRunner =
+    unsafe fn(DeferredWorld, observer: Entity, &TriggerContext, event: PtrMut, trigger: PtrMut);
 
-pub(super) fn observer_system_runner<E: Event, B: Bundle, S: ObserverSystem<E, B>>(
+/// # Safety
+///
+/// - `world` must be the [`DeferredWorld`] that the `entity` is defined in
+/// - `event_ptr` must match the `E` [`Event`] type.
+/// - `trigger_ptr` must match the [`Event::Trigger`] type for `E`.
+/// - `trigger_context`'s [`TriggerContext::event_key`] must match the `E` event type.
+///
+/// See the safety discussion on [`Trigger`] for more details.
```

Comment:
IMO that discussion would be much more appropriate here. The only relation it has with the `Trigger` trait is that triggers will most likely call this function and hence have to deal with its safety requirements (including the lifetime issue with `trigger_ptr`).